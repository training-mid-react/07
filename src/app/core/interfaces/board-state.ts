
export const ROWS = 6;
export const COLUMNS = 7;
export const WINNING_LENGTH = 4;

export interface IBoardGameState {
    board: BoardMatrix;
    currentPlayer:  Exclude<Player, 'draw'>;
    winner: Player;
}

export type BoardMatrix = Cell[][];
export type Cell =  Exclude<Player, 'draw'>;
export type Player = null | 'player1' | 'player2' | 'draw'
