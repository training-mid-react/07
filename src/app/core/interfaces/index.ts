export * from "./game"
export * from "./home"
export * from "./routes"
export * from "./state.interface"
export * from "./layoutBoard"