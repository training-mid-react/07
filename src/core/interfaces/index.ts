export enum PLAYER {
  PLAYER1 = 'player1',
  PLAYER2 = 'player2',
}
export type Cell = PLAYER | null;
export type IBoard = Cell[][];

export enum GAME_ACTIONS {
  DROP_TOKEN = 'DROP_TOKEN',
  RESET_GAME = 'RESET_GAME',
}

export type GameActionReducer =
  | { type: 'DROP_TOKEN'; columnIndex: number }
  | { type: 'RESET_GAME' };

export interface GameState {
  board: IBoard;
  currentPlayer: PLAYER;
  winner: PLAYER | null;
  isDraw: boolean;
  isDropping: boolean;
}

export interface GameProviderProps {
  children: React.ReactNode | React.ReactNode[];
}
