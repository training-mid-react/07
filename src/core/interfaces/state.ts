import { GameState } from ".";

// eslint-disable-next-line @typescript-eslint/no-empty-object-type
export interface IState extends GameState {
}

export interface IContext {
  state: IState;
  dispatch: React.Dispatch<IAction>;
}

export interface IAction {
  type: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  payload: any;
}