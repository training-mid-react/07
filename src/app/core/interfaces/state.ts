import { IBoardGameState } from './board-state';

export interface IContext {
  state: IState;
  dispatch: React.Dispatch<IAction>;
}

// eslint-disable-next-line @typescript-eslint/no-empty-object-type
export interface IState extends IBoardGameState {}

export interface IAction {
  type: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  payload?: any;
}