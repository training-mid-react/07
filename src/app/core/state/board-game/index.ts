import { COLUMNS, IBoardGameState, Player,  ROWS } from '../../interfaces/board-state';
import { BoardActionsType } from './actions';

// ==> Initial state
export const boardInitialState: IBoardGameState = {
  board: Array.from({ length: ROWS }, () => Array(COLUMNS).fill(null)),
  currentPlayer: 'player1',
  winner:  null,
};

// ===> Cases to modify state
export const boardCases =  {
  [BoardActionsType.ADD_PEAK] : (state: IBoardGameState, payload: Player[][]) => {
    return {
      ...state,
      board: payload
    };
  },

  [BoardActionsType.CHANGE_PLAYER] : (state: IBoardGameState, payload: Player) => {
    const newPlayer = payload ==='player1' ? 'player2' : 'player1';
    return {
      ...state,
      currentPlayer: newPlayer
    };
  },

  [BoardActionsType.SET_WINNER] : (state: IBoardGameState, payload: string) => {
    return {
      ...state,
      winner: payload
    };
  },

  [BoardActionsType.RESET_GAME]: (state: IBoardGameState) => {
    return {
      ...state,
      board: Array.from({ length: ROWS }, () => Array(COLUMNS).fill(null)),
      currentPlayer: 'player1',
      winner:  null,
    };
  },

};

