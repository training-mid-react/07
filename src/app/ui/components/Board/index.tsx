import { useState } from 'react';
import { BoardMatrix, Player } from '../../../core/interfaces/board-state';
import './style.scss';

interface Props{
  board: BoardMatrix,
  handleSelectColumn: (column: number) => void
  currentPlayer: Exclude<Player, 'draw'>
}
export const Board = ({board, handleSelectColumn, currentPlayer}: Props) => {
  
  const [animatedCell, setAnimatedCell] = useState<{ col: number; row: number } | null>(null);

  const handleCellClick = (colIndex: number) => {
    const rowIndex = board.findIndex(row => row[colIndex] === null);
    if (rowIndex !== -1) {
      setAnimatedCell({ col: colIndex, row: rowIndex });
      setTimeout(() => {
        handleSelectColumn(colIndex);
        setAnimatedCell(null);
      }, 500); 
    }
  };

  return (
    
    <div className="board">
      {
        board.map((row, rowIndex) => {
          return (
            <div 
              key={`row_${rowIndex}`} 
              className='board__row'
            >
              {row.map((cell, colIndex) => {
                const isFalling = animatedCell?.col === colIndex && animatedCell.row >= rowIndex;
                const fallDistance = isFalling ? `${50 * (animatedCell!.row - rowIndex)}px` : '0px';
                const animationFall = {
                  transform: `translateY(${fallDistance})`,
                };
                return( 
                  <div 
                    key={`cell_${colIndex}`} 
                    data-testid={`cell_${rowIndex}_${colIndex}`}
                    className={`board__col ${cell || ''} ${isFalling ? `falling--${currentPlayer}` : ''}`}  
                    onClick={() => handleCellClick(colIndex)}
                    style={isFalling ? animationFall : {}}
                  />
                );
              })}
            </div>
          );
        })
      }
      
    </div>
  );
};
