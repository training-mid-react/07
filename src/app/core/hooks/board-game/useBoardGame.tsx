import { ROWS } from '@core/constants';
import { AppContext } from '@core/state/AppContext';
import {
    setPlayeres,
    setWinner,
    updateBoard,
} from '@core/state/board-game/actions';
import { checkWinner } from '@core/utils';
import { useContext, useState } from 'react';

export const useBoardGame = () => {
    const { state, dispatch } = useContext(AppContext);
    const [fallingCells, setFallingCells] = useState<string[]>([]);

    const winner = state.winner;
    const board = state.board;
    const players = state.players;

    const handleClick = (col: number) => {
        if (winner) return;

        for (let row = ROWS - 1; row >= 0; row--) {
            if (board && !board[row][col]) {
                const newBoard = [...board];
                const currentPlayer = players?.find((p) => p.isCurrentPlayer);
                newBoard[row][col] = currentPlayer?.color ?? null;

                setFallingCells((prev) => [...prev, `${row}-${col}`]);
                dispatch(updateBoard(newBoard));

                const result = checkWinner({
                    board: newBoard,
                    row,
                    col,
                    players,
                });

                if (result) dispatch(setWinner(result));

                const [player1, player2] = players ?? [];
                dispatch(
                    setPlayeres([
                        {
                            ...player1,
                            isCurrentPlayer: !player1.isCurrentPlayer,
                        },
                        {
                            ...player2,
                            isCurrentPlayer: !player2.isCurrentPlayer,
                        },
                    ])
                );
                return;
            }
        }
    };

    return {
        players,
        winner,
        board,
        fallingCells,
        handleClick,
    };
};
