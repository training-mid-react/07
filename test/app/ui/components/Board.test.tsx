import React, { act } from 'react';
import { vi } from 'vitest';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';

import { Board } from '../../../../src/app/ui/components/Board';
import { boardInitialState } from '../../../../src/app/core/state/board-game';
import { COLUMNS, ROWS } from '../../../../src/app/core/interfaces/board-state';

describe('Board', () => {

  // ===> Initial state
  const initialState = Array.from({ length: ROWS }, () => Array(COLUMNS).fill(null));
  
  test('Matches the snapshot', () => {
    const initialState = boardInitialState.board;
    const handleSelectColumn = vi.fn();
    const component = render(
      <Board 
        board={initialState}
        currentPlayer={'player1'}
        handleSelectColumn={handleSelectColumn}
      />
    );
    expect(component.asFragment()).toMatchSnapshot();
  });

  test('Expect to put a player name in the table', async() => {
   
    const player = 'player1';
    
    const handleSelectColumn = vi.fn();            
    render(
      <Board 
        board={initialState}
        currentPlayer={player}
        handleSelectColumn={handleSelectColumn}
      />
    );

    const cell = screen.getByTestId('cell_5_0');
    await act(async() => {
      fireEvent.click(cell);
    });
    
    await waitFor(() => expect(handleSelectColumn).toHaveBeenCalledTimes(1));
  });

});