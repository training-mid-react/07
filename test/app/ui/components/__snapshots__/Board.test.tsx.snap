// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Board > Matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="board"
  >
    <div
      class="board__row"
    >
      <div
        class="board__col  "
        data-testid="cell_0_0"
      />
      <div
        class="board__col  "
        data-testid="cell_0_1"
      />
      <div
        class="board__col  "
        data-testid="cell_0_2"
      />
      <div
        class="board__col  "
        data-testid="cell_0_3"
      />
      <div
        class="board__col  "
        data-testid="cell_0_4"
      />
      <div
        class="board__col  "
        data-testid="cell_0_5"
      />
      <div
        class="board__col  "
        data-testid="cell_0_6"
      />
    </div>
    <div
      class="board__row"
    >
      <div
        class="board__col  "
        data-testid="cell_1_0"
      />
      <div
        class="board__col  "
        data-testid="cell_1_1"
      />
      <div
        class="board__col  "
        data-testid="cell_1_2"
      />
      <div
        class="board__col  "
        data-testid="cell_1_3"
      />
      <div
        class="board__col  "
        data-testid="cell_1_4"
      />
      <div
        class="board__col  "
        data-testid="cell_1_5"
      />
      <div
        class="board__col  "
        data-testid="cell_1_6"
      />
    </div>
    <div
      class="board__row"
    >
      <div
        class="board__col  "
        data-testid="cell_2_0"
      />
      <div
        class="board__col  "
        data-testid="cell_2_1"
      />
      <div
        class="board__col  "
        data-testid="cell_2_2"
      />
      <div
        class="board__col  "
        data-testid="cell_2_3"
      />
      <div
        class="board__col  "
        data-testid="cell_2_4"
      />
      <div
        class="board__col  "
        data-testid="cell_2_5"
      />
      <div
        class="board__col  "
        data-testid="cell_2_6"
      />
    </div>
    <div
      class="board__row"
    >
      <div
        class="board__col  "
        data-testid="cell_3_0"
      />
      <div
        class="board__col  "
        data-testid="cell_3_1"
      />
      <div
        class="board__col  "
        data-testid="cell_3_2"
      />
      <div
        class="board__col  "
        data-testid="cell_3_3"
      />
      <div
        class="board__col  "
        data-testid="cell_3_4"
      />
      <div
        class="board__col  "
        data-testid="cell_3_5"
      />
      <div
        class="board__col  "
        data-testid="cell_3_6"
      />
    </div>
    <div
      class="board__row"
    >
      <div
        class="board__col  "
        data-testid="cell_4_0"
      />
      <div
        class="board__col  "
        data-testid="cell_4_1"
      />
      <div
        class="board__col  "
        data-testid="cell_4_2"
      />
      <div
        class="board__col  "
        data-testid="cell_4_3"
      />
      <div
        class="board__col  "
        data-testid="cell_4_4"
      />
      <div
        class="board__col  "
        data-testid="cell_4_5"
      />
      <div
        class="board__col  "
        data-testid="cell_4_6"
      />
    </div>
    <div
      class="board__row"
    >
      <div
        class="board__col  "
        data-testid="cell_5_0"
      />
      <div
        class="board__col  "
        data-testid="cell_5_1"
      />
      <div
        class="board__col  "
        data-testid="cell_5_2"
      />
      <div
        class="board__col  "
        data-testid="cell_5_3"
      />
      <div
        class="board__col  "
        data-testid="cell_5_4"
      />
      <div
        class="board__col  "
        data-testid="cell_5_5"
      />
      <div
        class="board__col  "
        data-testid="cell_5_6"
      />
    </div>
  </div>
</DocumentFragment>
`;
